### ISSUE TITLE
pahole memory allocation problems

### ISSUE THREAD

**Issue by jeffmerkey** [acmel/dwarves#55]
When building RHEL 9.4 running as a qemu-kvm vm under gnome-boxes, it is required to configure a swap file size in excess of 10 GB for the kvm image to prevent pahole from crashing and triggering the OOM Killer and failing linux kernel build.  it requires excessive amounts of memory when parsing the BTF data from vmlinux.   gcc linking with ld works but when pahole executes it consumes excessive memory during 'make bzImage modules'.  It really needs to be rewritten with it's own internal lru or save data as intermediate files rather than just sucking up 6-7 GB of memory and crashing the system.  Is there some workaround for this issue?  Thanks.

**Comment by jeffmerkey** [comment]
I will look over the program and possibly suggest a fix for this.

**Comment by alan-maguire** [comment]
have you tested the "next" branch? It has multiple fixes around memory usage.

**Comment by jeffmerkey** [comment]
Hi,

I cloned the 'next' branch with:

git clone -b next --single-branch https://github.com/acmel/dwarves

I complied the pahole v1.27 program with the buildcmd.sh and moved it into a RHEL 9.4 /usr/bin and tested it.  The memory usage was about 2 GB total for a build of the 5.14 redhat kernel.  

This is an improvement over the pahole v1.25 version currently shipped with RHEL 9.4.  There still needs to be some improvement in memory usage overall.  It still requires a fairly large swap partition (over 4 GB) to prevent crashes.  I am guessing there is a memory leak for pahole v1.25 which was addressed in v1.27.  

So to summarize:

pahole v1.25 - requires 10 GB swap space to avoid OOM kill and crashing on RHEL 9.4.
pahole v1.27 - requires 4-6 GB swap space  to avoid OOM kill and crashing on RHEL 9.4.

Using intermediate tmp files like git does for storing data to disk so memory use is not excessive for pahole.
